### ISSUE TITLE
Proposal: JavascriptException combined original and javascript stack traces

### ISSUE THREAD

**Issue by christianrondeau** [sebastienros/jint#1164]
I often have errors such as this:

```
Jint.Runtime.JavaScriptException: Error thrown in js
   at MyClass () my-module:5:9
   at my-module:6:16
```

The problem is that I cannot say where _in the C# code_ this actually happened.

## Variation 1: Combined stack traces

What I suggest (as a first draft to validate if there's interest) would be to combine both stack traces into one, like this:

```
Jint.Runtime.JavaScriptException: Error thrown in js
   at MyClass () my-module:5:9
   at my-module:6:16
   at MyProject.MyJavascriptRunner.Run() in C:\...\MyJavascriptRunner.cs:line 43
```

My initial implementation idea would be to change JavaScriptException.StackTrace to return both the base stack trace and the js stack trace, however that could be weird.

## Variation 2: Inner exception

An alternative implementation is two create two exceptions:

```
Jint.Runtime.JavaScriptException: Error thrown in js
   at MyClass () my-module:5:9
   at my-module:6:16
Jint.Runtime.JavaScriptExecutionException: An error was thrown by the script
   at MyProject.MyJavascriptRunner.Run() in C:\...\MyJavascriptRunner.cs:line 43
```

1. Keep the JavaScriptException stack trace as-is
2. Wrap (inner exception) in a JavaScriptExecutionException that would contain the "real" stack trace
3. Add property called JavaScriptStackTrace to JavaScriptExecutionException to make it easier to get a user stack trace

## Formatting

Finally, I think the formatting is currently different than the normal C# stack trace formatting, I would suggest we use C# formatting for both.

Before I spend more time investigating this, is there any interest? That would be a breaking change.

**Comment by lahma** [comment]
I'm haven't done much digging, but some thoughts/remarks:

* stack traces and errors should follow node/common browser formatting
* we cannot expose c# exception stacks by default, they "shouldn't happen" and are host problems and might cause security concerns, should be behind a switch (like in ASP.NET development whether to expose actual error details)
* there's `Error.cause` already implemented IIRC, https://github.com/tc39/proposal-error-cause
   * `AggregateError` is also there, but I guess it's for different purposes, and it hasn't been implemented (yet)

**Comment by Genteure** [comment]
I think the main problem is the .NET call stack is not available on `JavaScriptException` so it makes debugging the host application a bit difficult.

You can still "recover" the original .NET call stack by using `new System.Diagnostics.StackTrace(exception)`

```csharp
try
{
    engine.Execute(@"throw new Error('test')");
}
catch (Exception e)
{
    var s = new StackTrace(e);
    
    e.ToString();

    // Jint.Runtime.JavaScriptException: test
    //    at <anonymous>:7:1

    s.ToString();
    
    //    at Jint.Engine.ScriptEvaluation(ScriptRecord scriptRecord)
    //    at Jint.Engine.<>c__DisplayClass63_0.<Execute>b__0()
    //    at Jint.Engine.ExecuteWithConstraints[T](Boolean strict, Func`1 callback)
    //    at Jint.Engine.Execute(Script script)
    //    at Jint.Engine.Execute(String source, ParserOptions parserOptions)
    //    at Jint.Engine.Execute(String source)
    //    at JintExperiments.Program.Main(String[] args)
}
```

I think there are 3 different use cases that would prefer different call stacks:

## Debugging the script

In this case I would agree:

> - stack traces and errors should follow node/common browser formatting
> - we cannot expose c# exception stacks by default

## Debugging Jint

If I'm trying find if there's a problem with Jint, I would want to have the full .NET call stack.
JS call stack would be helpful but not necessarily required.

## Debugging the host application

Combined call stack would be best for this case, it would provide info about the interaction between the host application and the script at a glance.

We could have something slimiar to Visual Studio's "Call Stack" window with native code debugging enabled.

> ![](https://i.stack.imgur.com/Gbt8W.png)
> _Image from [this StackOverflow answer](https://stackoverflow.com/a/15662056)_

I haven't looked into the code so I don't know if this would be too much work.
A sperated JS stack and a .NET stack would also be sufficient.

If anyone want to make significant changes around call stacks, please also consider make the current call stack available.
My use case is exactly the same as #1115 -- to implement `console`.

**Comment by christianrondeau** [comment]
Thanks @Genteure I meant to answer this too :) My initial issue @lahma was getting the C# stack trace, AND being able to figure out what was wrong in the JavaScript code itself.

For example, this would only return JS-formatted stack traces of JS code:

```js
// In javascript:
try {
  throw new Error('test');
} catch (e) {
  // Here, e would never contain C# stack traces
  / and would return js-formatted stacks.
}
```

However, this needs the C# stack trace:

```c#
// In c#:
try {
  engine.Evaluate("throw new Error('test');")
} catch (Exception e) {
  // Here, I expect e to contain the C# stack trace
  / and use C# formatting
}
```

My expectation would be that _only_ the C# stack trace would be present in a C# exception; if I need the JS stack trace, I'd catch JavaScriptException and get a property like ScriptStackTrace for example.

However, I think it's nice to have both the JS stack trace _and_ the C# stack trace together and it makes sense; but that would be using C# stack trace formatting IMHO.

So, TL;DR is (my opinion), C# exceptions must have C# stacks, JS errors must have JS stacks, and if there is consensus, C# exceptions could have both.

What do you think?

**Comment by lahma** [comment]
> So, TL;DR is (my opinion), C# exceptions must have C# stacks, JS errors must have JS stacks, and if there is consensus, C# exceptions could have both.

This sounds correct to me 👍🏻

**Comment by christianrondeau** [comment]
Cool, I might try my hand at a draft PR sometime next week, the general idea I have in mind is to:
1. Return the JavaScript stack trace in a "JavaScriptStackTrace" property of the exception, in the same format as Node;
2. Overload StackTrace to return both the C# stack trace and the JavaScript stack tracem using the same format as C# for both;
3. Add an OriginalStackTrace property that will return the unmodified C# stack trace of the exception.

**Comment by christianrondeau** [comment]
I pushed a draft, I'm not sure about a few things, opinions welcome before I spend time on the wrong things :)

**Comment by christianrondeau** [comment]
Closed through #1171
